src/mender-update/standalone/states.cpp

```diff
@@ -0,0 +1,507 @@
+
+	err = ctx.update_module->CleanAndPrepareFileTree(
+		ctx.update_module->GetUpdateModuleWorkDir(), header);
+	if (err != error::NoError) {
+		UpdateResult(ctx.result_and_error, {Result::Failed | Result::NoRollbackNecessary, err});
+		poster.PostEvent(StateEvent::Failure);
+		return;
+	}
+
+	ctx.state_data = StateDataFromPayloadHeaderView(header);
+
+	if (ctx.options != InstallOptions::NoStdout) {
+		cout << "Installing artifact..." << endl;
+	}
+
+	auto exp_matches = main_context.MatchesArtifactDepends(header.header);
+	if (!exp_matches) {
+		UpdateResult(
+			ctx.result_and_error,
+			{Result::Failed | Result::NoRollbackNecessary, exp_matches.error()});
+		poster.PostEvent(StateEvent::Failure);
+		return;
+	} else if (!exp_matches.value()) {
+		// reasons already logged
+		UpdateResult(
+			ctx.result_and_error, {Result::Failed | Result::NoRollbackNecessary, error::NoError});
+		poster.PostEvent(StateEvent::Failure);
+		return;
+	}
+
+	poster.PostEvent(StateEvent::Success);
+}
+
+void DownloadState::OnEnter(Context &ctx, sm::EventPoster<StateEvent> &poster) {
+	auto err = DoDownloadState(ctx);
+	if (err != error::NoError) {
+		log::Error("Streaming failed: " + err.String());
+		UpdateResult(ctx.result_and_error, {Result::Failed | Result::NoRollbackNecessary, err});
+		poster.PostEvent(StateEvent::Failure);
+		return;
+	}
+
+	UpdateResult(ctx.result_and_error, {Result::Downloaded, error::NoError});
+	poster.PostEvent(StateEvent::Success);
+}
+
+void ArtifactInstallState::OnEnter(Context &ctx, sm::EventPoster<StateEvent> &poster) {
+	auto err = ctx.update_module->ArtifactInstall();
+	if (err != error::NoError) {
+		log::Error("Installation failed: " + err.String());
+		UpdateResult(ctx.result_and_error, {Result::Failed, err});
+		poster.PostEvent(StateEvent::Failure);
+		return;
+	}
+
+	UpdateResult(ctx.result_and_error, {Result::Installed, error::NoError});
+	poster.PostEvent(StateEvent::Success);
+}
+
+void RebootAndRollbackQueryState::OnEnter(Context &ctx, sm::EventPoster<StateEvent> &poster) {
+	auto reboot = ctx.update_module->NeedsReboot();
+	if (!reboot) {
+		log::Error("Could not query for reboot: " + reboot.error().String());
+		UpdateResult(ctx.result_and_error, {Result::Failed, reboot.error()});
+		poster.PostEvent(StateEvent::Failure);
+		return;
+	}
+
+	if (reboot.value() != update_module::RebootAction::No) {
+		UpdateResult(ctx.result_and_error, {Result::RebootRequired, error::NoError});
+	}
+
+	auto rollback_support = ctx.update_module->SupportsRollback();
+	if (!rollback_support) {
+		log::Error("Could not query for rollback support: " + rollback_support.error().String());
+		UpdateResult(ctx.result_and_error, {Result::Failed, rollback_support.error()});
+		poster.PostEvent(StateEvent::Failure);
+		return;
+	}
+
+	if (rollback_support.value()) {
+		poster.PostEvent(StateEvent::NeedsInteraction);
+		return;
+	}
+
+	cout << "Update Module doesn't support rollback. Committing immediately." << endl;
```

Comment:
It does, actually. This is a big difference between standalone mode and daemon mode. Standalone mode doesn't deal with rebooting at all, so it just commits immediately if there is no chance of rollback. In hindsight, maybe this was a bad decision, since `ArtifactCommit_Leave` script may actually depend on running a new update. But the Golang client also does this, so we won't change this now. And it only affects update modules without rollback, so rootfs-image updates are not affected.

Reply:
Thanks for clarifying this :memo: